File: ticru-cli.py
#!/usr/bin/env python3
"""
Ticru.io CLI Tool
Command-line interface for managing the Ticru.io application
"""

import click
import subprocess
import os
import sys
import signal
from pathlib import Path
from multiprocessing import Process

@click.group()
@click.version_option(version='1.0.0')
def cli():
    """Ticru.io Command Line Interface"""
    pass

@cli.command()
@click.option('--port', default=8000, help='Port to run the server on')
@click.option('--host', default='0.0.0.0', help='Host to bind to')
def serve(port, host):
    """Start the API server"""
    click.echo(f"🚀 Starting Ticru.io API server on {host}:{port}")
    subprocess.run([
        'uvicorn',
        'api-server:app',
        '--host', host,
        '--port', str(port),
        '--reload'
    ])

@cli.command()
@click.option('--port', default=5173, help='Port to run the dev server on')
def dev(port):
    """Start the development server"""
    click.echo(f"🔧 Starting development server on port {port}")
    subprocess.run(['npm', 'run', 'dev', '--', '--port', str(port)])

@cli.command()
@click.option('--frontend-port', default=5173, help='Port for the frontend dev server')
@click.option('--backend-port', default=8000, help='Port for the backend API server')
@click.option('--backend-host', default='0.0.0.0', help='Host for the backend server')
def run(frontend_port, backend_port, backend_host):
    """Start both frontend and backend servers concurrently"""
    click.echo("🚀 Starting Ticru.io Application")
    click.echo("="*60)
    click.echo(f"  Frontend Dev Server: http://localhost:{frontend_port}")
    click.echo(f"  Backend API Server:  http://{backend_host}:{backend_port}")
    click.echo("="*60)
    click.echo("\nPress Ctrl+C to stop both servers\n")
    
    # Start backend server
    def start_backend():
        subprocess.run([
            'uvicorn',
            'api-server:app',
            '--host', backend_host,
            '--port', str(backend_port),
            '--reload'
        ])
    
    # Start frontend server
    def start_frontend():
        subprocess.run(['npm', 'run', 'dev', '--', '--port', str(frontend_port)])
    
    # Create processes
    backend_process = Process(target=start_backend)
    frontend_process = Process(target=start_frontend)
    
    # Handle graceful shutdown
    def signal_handler(sig, frame):
        click.echo("\n\n🛑 Shutting down servers...")
        backend_process.terminate()
        frontend_process.terminate()
        backend_process.join(timeout=5)
        frontend_process.join(timeout=5)
        click.echo("✅ Servers stopped")
        sys.exit(0)
    
    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)
    
    try:
        # Start both servers
        backend_process.start()
        frontend_process.start()
        
        # Wait for processes
        backend_process.join()
        frontend_process.join()
    except KeyboardInterrupt:
        click.echo("\n\n🛑 Shutting down servers...")
        backend_process.terminate()
        frontend_process.terminate()
        backend_process.join(timeout=5)
        frontend_process.join(timeout=5)
        click.echo("✅ Servers stopped")

@cli.command()
def build():
    """Build the application for production"""
    click.echo("🏗️  Building application...")
    subprocess.run(['python3', 'build-system.py', '--all'])

@cli.command()
def install():
    """Install all dependencies"""
    click.echo("📦 Installing dependencies...")
    click.echo("\n→ Installing npm dependencies...")
    subprocess.run(['npm', 'install'])
    
    if Path('requirements.txt').exists():
        click.echo("\n→ Installing Python dependencies...")
        subprocess.run(['pip', 'install', '-r', 'requirements.txt'])
    
    click.echo("\n✅ Dependencies installed successfully!")

@cli.command()
def test():
    """Run all tests"""
    click.echo("🧪 Running tests...")
    click.echo("⚠️  Tests not configured yet")

@cli.command()
def lint():
    """Run code linting"""
    click.echo("🔍 Running linters...")
    subprocess.run(['npm', 'run', 'lint'])

@cli.command()
@click.option('--sql-file', default='BUILD-DATABASE.sql', help='SQL file to execute')
def init_db(sql_file):
    """Initialize the database"""
    click.echo(f"🗄️  Initializing database from {sql_file}")
    
    if not Path(sql_file).exists():
        click.echo(f"❌ Error: {sql_file} not found", err=True)
        sys.exit(1)
    
    db_url = os.getenv('DATABASE_URL', 'postgresql://localhost/ticru_db')
    click.echo(f"Database URL: {db_url}")
    
    try:
        subprocess.run([
            'psql',
            db_url,
            '-f',
            sql_file
        ], check=True)
        click.echo("✅ Database initialized successfully!")
    except subprocess.CalledProcessError:
        click.echo("❌ Failed to initialize database", err=True)
        click.echo("Make sure PostgreSQL is running and DATABASE_URL is set correctly")
        sys.exit(1)

@cli.command()
def clean():
    """Clean build artifacts"""
    click.echo("🧹 Cleaning build artifacts...")
    
    paths = ['dist', 'build', 'node_modules/.cache', '__pycache__', '*.pyc']
    
    for path in paths:
        if '*' in path:
            subprocess.run(['find', '.', '-name', path, '-delete'], check=False)
        elif Path(path).exists():
            subprocess.run(['rm', '-rf', path])
            click.echo(f"  Removed {path}")
    
    click.echo("✅ Cleanup complete!")

@cli.command()
@click.option('--platform', type=click.Choice(['vercel', 'netlify']), required=True)
def deploy(platform):
    """Deploy to production"""
    click.echo(f"🚀 Deploying to {platform}...")
    
    if platform == 'vercel':
        if not Path('deploy-vercel.sh').exists():
            click.echo("❌ deploy-vercel.sh not found", err=True)
            sys.exit(1)
        subprocess.run(['bash', 'deploy-vercel.sh'])
    elif platform == 'netlify':
        subprocess.run(['netlify', 'deploy', '--prod'])
    
    click.echo("✅ Deployment complete!")

@cli.command()
def setup():
    """Run the complete setup wizard"""
    click.echo("🎯 Ticru.io Setup Wizard")
    click.echo("="*50)
    
    # Check Node.js
    click.echo("\n1. Checking Node.js...")
    result = subprocess.run(['node', '--version'], capture_output=True, text=True)
    if result.returncode == 0:
        click.echo(f"   ✅ Node.js {result.stdout.strip()}")
    else:
        click.echo("   ❌ Node.js not found")
        sys.exit(1)
    
    # Check Python
    click.echo("\n2. Checking Python...")
    result = subprocess.run(['python3', '--version'], capture_output=True, text=True)
    if result.returncode == 0:
        click.echo(f"   ✅ Python {result.stdout.strip()}")
    else:
        click.echo("   ❌ Python not found")
        sys.exit(1)
    
    # Install dependencies
    click.echo("\n3. Installing dependencies...")
    subprocess.run(['npm', 'install'])
    if Path('requirements.txt').exists():
        subprocess.run(['pip', 'install', '-r', 'requirements.txt'])
    click.echo("   ✅ Dependencies installed")
    
    # Create .env if not exists
    click.echo("\n4. Setting up environment...")
    if not Path('.env').exists() and Path('.env.example').exists():
        subprocess.run(['cp', '.env.example', '.env'])
        click.echo("   ✅ Created .env file")
    else:
        click.echo("   ✅ Environment already configured")
    
    click.echo("\n" + "="*50)
    click.echo("✅ Setup complete!")
    click.echo("\nNext steps:")
    click.echo("  • Run 'python3 ticru-cli.py dev' to start development server")
    click.echo("  • Run 'python3 ticru-cli.py serve' to start API server")
    click.echo("  • Run 'python3 ticru-cli.py build' to build for production")

@cli.command()
def status():
    """Show application status"""
    click.echo("📊 Ticru.io Status")
    click.echo("="*50)
    
    # Check if servers are running
    click.echo("\nServer Status:")
    
    # Check API server
    try:
        import requests
        response = requests.get('http://localhost:8000/api/health', timeout=2)
        if response.status_code == 200:
            click.echo("  API Server: ✅ Running (port 8000)")
        else:
            click.echo("  API Server: ⚠️  Unhealthy")
    except:
        click.echo("  API Server: ❌ Not running")
    
    # Check dev server
    try:
        import requests
        response = requests.get('http://localhost:5173', timeout=2)
        if response.status_code == 200:
            click.echo("  Dev Server: ✅ Running (port 5173)")
        else:
            click.echo("  Dev Server: ⚠️  Unhealthy")
    except:
        click.echo("  Dev Server: ❌ Not running")
    
    # Build status
    click.echo("\nBuild Status:")
    if Path('dist').exists():
        files = list(Path('dist').rglob('*'))
        click.echo(f"  Build Output: ✅ {len(files)} files")
    else:
        click.echo("  Build Output: ❌ Not built")
    
    # Dependencies
    click.echo("\nDependencies:")
    if Path('node_modules').exists():
        click.echo("  NPM Packages: ✅ Installed")
    else:
        click.echo("  NPM Packages: ❌ Not installed")

if __name__ == '__main__':
    cli()

File: ticru-cli.ts
#!/usr/bin/env node

/**
 * Ticru.io CLI Tool
 * Command-line interface for managing the Ticru.io application
 */

import { Command } from 'commander';
import { spawn, ChildProcess } from 'child_process';
import { config } from 'dotenv';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';
import { readFileSync, existsSync, rmSync } from 'fs';
import { execSync } from 'child_process';

// Load environment variables
config();

const program = new Command();

// Get current directory
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Package version
const packageJson = JSON.parse(
  readFileSync(join(__dirname, 'package.json'), 'utf-8')
);

program
  .name('ticru-cli')
  .description('Ticru.io Command Line Interface')
  .version(packageJson.version || '1.0.0');

/**
 * Start the API server
 */
program
  .command('serve')
  .description('Start the API server')
  .option('-p, --port <port>', 'Port to run the server on', '8000')
  .option('-h, --host <host>', 'Host to bind to', '0.0.0.0')
  .action((options) => {
    console.log(`🚀 Starting Ticru.io API server on ${options.host}:${options.port}`);

    const env = { ...process.env, PORT: options.port, HOST: options.host };

    spawn('npm', ['run', 'dev:api'], {
      stdio: 'inherit',
      shell: true,
      env,
    });
  });

/**
 * Start the development server
 */
program
  .command('dev')
  .description('Start the development server')
  .option('-p, --port <port>', 'Port to run the dev server on', '5173')
  .action((options) => {
    console.log(`🔧 Starting development server on port ${options.port}`);

    spawn('npm', ['run', 'dev', '--', '--port', options.port], {
      stdio: 'inherit',
      shell: true,
    });
  });

/**
 * Start both frontend and backend servers concurrently
 */
program
  .command('run')
  .description('Start both frontend and backend servers concurrently')
  .option('-fp, --frontend-port <port>', 'Port for the frontend dev server', '5173')
  .option('-bp, --backend-port <port>', 'Port for the backend API server', '8000')
  .option('-bh, --backend-host <host>', 'Host for the backend server', '0.0.0.0')
  .action((options) => {
    console.log('🚀 Starting Ticru.io Application');
    console.log('='.repeat(60));
    console.log(`  Frontend Dev Server: http://localhost:${options.frontendPort}`);
    console.log(`  Backend API Server:  http://${options.backendHost}:${options.backendPort}`);
    console.log('='.repeat(60));
    console.log('\nPress Ctrl+C to stop both servers\n');

    const processes: ChildProcess[] = [];

    // Start backend server
    const backendEnv = {
      ...process.env,
      PORT: options.backendPort,
      HOST: options.backendHost,
    };

    const backend = spawn('npm', ['run', 'dev:api'], {
      stdio: 'inherit',
      shell: true,
      env: backendEnv,
    });
    processes.push(backend);

    // Start frontend server
    const frontend = spawn('npm', ['run', 'dev', '--', '--port', options.frontendPort], {
      stdio: 'inherit',
      shell: true,
    });
    processes.push(frontend);

    // Handle graceful shutdown
    const shutdown = () => {
      console.log('\n\n🛑 Shutting down servers...');
      processes.forEach((proc) => {
        if (proc && !proc.killed) {
          proc.kill('SIGTERM');
        }
      });

      setTimeout(() => {
        processes.forEach((proc) => {
          if (proc && !proc.killed) {
            proc.kill('SIGKILL');
          }
        });
        console.log('✅ Servers stopped');
        process.exit(0);
      }, 5000);
    };

    process.on('SIGINT', shutdown);
    process.on('SIGTERM', shutdown);
  });

/**
 * Build the application for production
 */
program
  .command('build')
  .description('Build the application for production')
  .action(() => {
    console.log('🏗️  Building application...');
    try {
      execSync('node build-system.ts --all', { stdio: 'inherit' });
    } catch (error) {
      console.error('❌ Build failed');
      process.exit(1);
    }
  });

/**
 * Install all dependencies
 */
program
  .command('install')
  .description('Install all dependencies')
  .action(() => {
    console.log('📦 Installing dependencies...');
    console.log('\n→ Installing npm dependencies...');
    
    try {
      execSync('npm install', { stdio: 'inherit' });
      console.log('\n✅ Dependencies installed successfully!');
    } catch (error) {
      console.error('❌ Failed to install dependencies');
      process.exit(1);
    }
  });

/**
 * Run all tests
 */
program
  .command('test')
  .description('Run all tests')
  .action(() => {
    console.log('🧪 Running tests...');
    console.log('⚠️  Tests not configured yet');
  });

/**
 * Run code linting
 */
program
  .command('lint')
  .description('Run code linting')
  .action(() => {
    console.log('🔍 Running linters...');
    try {
      execSync('npm run lint', { stdio: 'inherit' });
    } catch (error) {
      console.error('❌ Linting failed');
      process.exit(1);
    }
  });

/**
 * Initialize the database
 */
program
  .command('init-db')
  .description('Initialize the database')
  .option('-f, --sql-file <file>', 'SQL file to execute', 'BUILD-DATABASE.sql')
  .action((options) => {
    console.log(`🗄️  Initializing database from ${options.sqlFile}`);

    if (!existsSync(options.sqlFile)) {
      console.error(`❌ Error: ${options.sqlFile} not found`);
      process.exit(1);
    }

    const dbUrl = process.env.DATABASE_URL || 'postgresql://localhost/ticru_db';
    console.log(`Database URL: ${dbUrl}`);

    try {
      execSync(`psql ${dbUrl} -f ${options.sqlFile}`, { stdio: 'inherit' });
      console.log('✅ Database initialized successfully!');
    } catch (error) {
      console.error('❌ Failed to initialize database');
      console.error('Make sure PostgreSQL is running and DATABASE_URL is set correctly');
      process.exit(1);
    }
  });

/**
 * Clean build artifacts
 */
program
  .command('clean')
  .description('Clean build artifacts')
  .action(() => {
    console.log('🧹 Cleaning build artifacts...');

    const paths = ['dist', 'build', 'node_modules/.cache'];

    paths.forEach((path) => {
      if (existsSync(path)) {
        rmSync(path, { recursive: true, force: true });
        console.log(`  Removed ${path}`);
      }
    });

    console.log('✅ Cleanup complete!');
  });

/**
 * Deploy to production
 */
program
  .command('deploy')
  .description('Deploy to production')
  .requiredOption('-p, --platform <platform>', 'Deployment platform (vercel or netlify)')
  .action((options) => {
    console.log(`🚀 Deploying to ${options.platform}...`);

    if (options.platform === 'vercel') {
      if (!existsSync('deploy-vercel.sh')) {
        console.error('❌ deploy-vercel.sh not found');
        process.exit(1);
      }
      try {
        execSync('bash deploy-vercel.sh', { stdio: 'inherit' });
        console.log('✅ Deployment complete!');
      } catch (error) {
        console.error('❌ Deployment failed');
        process.exit(1);
      }
    } else if (options.platform === 'netlify') {
      try {
        execSync('netlify deploy --prod', { stdio: 'inherit' });
        console.log('✅ Deployment complete!');
      } catch (error) {
        console.error('❌ Deployment failed');
        process.exit(1);
      }
    } else {
      console.error(`❌ Unknown platform: ${options.platform}`);
      process.exit(1);
    }
  });

/**
 * Run the complete setup wizard
 */
program
  .command('setup')
  .description('Run the complete setup wizard')
  .action(() => {
    console.log('🎯 Ticru.io Setup Wizard');
    console.log('='.repeat(50));

    // Check Node.js
    console.log('\n1. Checking Node.js...');
    try {
      const nodeVersion = execSync('node --version', { encoding: 'utf-8' }).trim();
      console.log(`   ✅ Node.js ${nodeVersion}`);
    } catch (error) {
      console.log('   ❌ Node.js not found');
      process.exit(1);
    }

    // Install dependencies
    console.log('\n2. Installing dependencies...');
    try {
      execSync('npm install', { stdio: 'inherit' });
      console.log('   ✅ Dependencies installed');
    } catch (error) {
      console.error('   ❌ Failed to install dependencies');
      process.exit(1);
    }

    // Create .env if not exists
    console.log('\n3. Setting up environment...');
    if (!existsSync('.env') && existsSync('.env.example')) {
      try {
        const envExample = readFileSync('.env.example', 'utf-8');
        require('fs').writeFileSync('.env', envExample);
        console.log('   ✅ Created .env file');
        console.log('   ⚠️  Please edit .env and add your configuration');
      } catch (error) {
        console.error('   ❌ Failed to create .env file');
      }
    } else {
      console.log('   ✅ Environment already configured');
    }

    console.log('\n' + '='.repeat(50));
    console.log('✅ Setup complete!');
    console.log('\nNext steps:');
    console.log('  • Run \'npm run cli dev\' to start development server');
    console.log('  • Run \'npm run cli serve\' to start API server');
    console.log('  • Run \'npm run cli build\' to build for production');
  });

/**
 * Show application status
 */
program
  .command('status')
  .description('Show application status')
  .action(async () => {
    console.log('📊 Ticru.io Status');
    console.log('='.repeat(50));

    console.log('\nServer Status:');

    // Check API server
    try {
      const response = await fetch('http://localhost:8000/api/health', {
        signal: AbortSignal.timeout(2000),
      });
      if (response.ok) {
        console.log('  API Server: ✅ Running (port 8000)');
      } else {
        console.log('  API Server: ⚠️  Unhealthy');
      }
    } catch (error) {
      console.log('  API Server: ❌ Not running');
    }

    // Check dev server
    try {
      const response = await fetch('http://localhost:5173', {
        signal: AbortSignal.timeout(2000),
      });
      if (response.ok) {
        console.log('  Dev Server: ✅ Running (port 5173)');
      } else {
        console.log('  Dev Server: ⚠️  Unhealthy');
      }
    } catch (error) {
      console.log('  Dev Server: ❌ Not running');
    }

    // Build status
    console.log('\nBuild Status:');
    if (existsSync('dist')) {
      const files = require('fs').readdirSync('dist', { recursive: true });
      console.log(`  Build Output: ✅ ${files.length} files`);
    } else {
      console.log('  Build Output: ❌ Not built');
    }

    // Dependencies
    console.log('\nDependencies:');
    if (existsSync('node_modules')) {
      console.log('  NPM Packages: ✅ Installed');
    } else {
      console.log('  NPM Packages: ❌ Not installed');
    }
  });

program.parse();
